src/main/java/io/cryostat/core/agent/LocalProbeTemplateService.java

```diff
@@ -0,0 +1,157 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package io.cryostat.core.agent;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import io.cryostat.core.FlightRecorderException;
+import io.cryostat.core.sys.Environment;
+import io.cryostat.core.sys.FileSystem;
+
+import org.xml.sax.SAXException;
+
+public class LocalProbeTemplateService extends AbstractProbeTemplateService {
+
+    public static final String TEMPLATE_PATH = "CRYOSTAT_PROBE_TEMPLATE_PATH";
+
+    private final FileSystem fs;
+    private final Environment env;
+
+    public LocalProbeTemplateService(FileSystem fs, Environment env) throws IOException {
+        this.fs = fs;
+        this.env = env;
+        // Sanity check
+        if (!env.hasEnv(TEMPLATE_PATH)) {
+            throw new IOException(
+                    String.format(
+                            "Probe template directory does not exist, must be set using environment variable %s",
+                            TEMPLATE_PATH));
+        }
+        // Sanity check the directory is set up correctly
+        Path probeTemplateDirectory = fs.pathOf(env.getEnv(TEMPLATE_PATH));
+        if (!fs.exists(probeTemplateDirectory)
+                || !fs.isDirectory(probeTemplateDirectory)
+                || !fs.isReadable(probeTemplateDirectory)
+                || !fs.isWritable(probeTemplateDirectory)) {
+            throw new IOException(
+                    String.format(
+                            "Probe template directory %s does not exist, is not a directory, or has incorrect permissions.",
+                            probeTemplateDirectory.toString()));
+        }
+    }
+
+    public void addTemplate(InputStream inputStream, String filename) throws Exception {
+        try (inputStream) {
+            ProbeTemplate template = new ProbeTemplate();
+            // If validation fails this will throw a ProbeValidationException with details
+            template.deserialize(inputStream);
+            Path path = fs.pathOf(env.getEnv(TEMPLATE_PATH), filename);
+            if (fs.exists(path)) {
+                throw new Exception(
+                        String.format(
+                                "Event template \"%s\" already exists", template.getFileName()));
+            }
+            fs.writeString(
+                    path,
+                    new BufferedReader(new InputStreamReader(inputStream, StandardCharsets.UTF_8))
+                            .lines()
+                            .collect(Collectors.joining("\n")));
+        } catch (Exception e) {
+            // rethrow for http handler
+            throw e;
```

Comment:
This `catch` clause can just be removed, since the method already declares `throws Exception`